20240912-R-scheduler-auto-scale.md

```diff
@@ -0,0 +1,110 @@
+# Auto-Scale the Scheduler service
+
+* Author(s): Cassie Coyle (@cicoyle)
+* State: Review
+* Updated: 2024-09-12
+
+## Overview
+
+This proposal outlines a design for auto-scaling the Schedulers, which is needed for the Scheduler service to be Stable.
+The Scheduler's internal cron scheduler library code will need to be refactored to be dynamic. Runtime will need to be
+updated to also account for dynamically connecting to Schedulers when they are scaled up or down. The Scheduler server
+code will need to be updated such that it optionally runs embedded etcd on different Scheduler instances.
+
+
+## Background
+
+In Dapr v1.14 we introduced the Scheduler control plane service to support the Jobs API along with scaling the Actor
+reminder system and the Workflow API (since it uses Actor reminders under the hood). The Scheduler encompasses running
+itself as a new process, an embedded etcd instance, and an internal cron scheduler. There are a few things needed to
+make the Scheduler a stable control plane service. This proposal will be detailing how to support the auto-scaling of
+Schedulers and secondly enhancing the Schedulers such that we optionally run etcd instances in the Schedulers under
+certain circumstances that make sense. When scaling beyond 5 etcd instances, there are little performance improvements, 
+so it makes sense for Schedulers to become stateless at this point. For example, we could have 20 Schedulers, but only
+5 of them would contain the embedded etcd.
+
+## Related Items
+
+### Related issues
+
+[Tracking Issue: Scheduler Stable](https://github.com/dapr/dapr/issues/7796) &&
+[Auto Scale Scheduler Issue](https://github.com/dapr/dapr/issues/7912).
+
+## Expectations and alternatives
+
+There will be 3 changes needed for this feature to be complete. 
+
+1. **New control loop** for the [go-etcd-cron library](https://github.com/diagridio/go-etcd-cron), such that the leadership table is able to be dynamic. This 
```

Comment:
host address does not mean much as someone can run multiple scheduler instances on the same host. Why we need host address?

Reply:
Host address should include port here.

@cicoyle can we include the proto of the leadership table value we talked about? I think it was something like the following.

```proto
message Leader {
  uint32 observed_total = 1;
  string host = 2;
  uint32 port = 3;
}

Reply:
Ah, it's written further down 🤠

Reply:
Using the host address and port is probably fairly brittle. For example, a controller (such as kubernetes) may move the workload to a different container on another machine. However, the 'node' will come back up with the same data but a different host + port combination.

From experience (running typesense in production on kubernetes), this results in a pretty brittle situation.

Instead, it is probably better to accept a "name" parameter and fallback to the hostname. In this way, it will "just work" with stateful sets in kubernetes (which sets the hostname as $POD-$NUM). Further it allows for running on the same host by setting 'name' to unique values.

Reply:
As a counter argument, I don’t think this is brittle as the stored and broadcast addresses are not used as any type of identifier, but instead as just routable addresses the client should connect to. Clients do not care which scheduler is what, just so long as it connects to all those that are available.

When a scheduler workload moves to another host address, the new address should be either updated with the new one (ip), or clients will re-resolve the host name to obtain the new ip (kube dns). When a Sched workload moves, all clients would have disconnected naturally, and begin the process of re-resolving or receive an update from the `WatchHosts` stream.

We prevent stale addresses from being persisted in the leadership table by an optimistic delete on Sched shutdown, and an Etcd key lease.

What we are trying to achieve here is the client needing only a single main Scheduler host address to connect to, and then discover the rest. This main address would ideally be a resolved service cluster ip that picks any Scheudler in the set, reresolving on disconnect. In self hosted, it would be the first scheduled in the set- which would always be the case at least.

Reply:
I understand what you are saying, but that isn't how etcd/raft works; it is worth being explicit in the document. In any case, there is a failure condition in your elucidation, best shown by example:

1. a physical machine goes down or node gets disconnected
2. the lease will still be active AND in the leadership table
3. eventually the lease will be released, but
4. when the container is moved, it will come back, and if you previously had three etcd nodes, now you will have four.

It is more robust to design for the failure case and treat the failure case as the normal case. In other words, with hostname (or generic name), you end up with a static set in etcd. The address of the set is an implementation detail (defined through some 'well-known name', such as 
a cluster-ip, dns name, or whatever). If a node disappears from the network, it will eventually come back (or at least another one with the same name that can catch up). This allows etcd to do what it does best, while we focus on scheduling. There is no need for leases, leadership tables, etc.

The main issue with the proposed solution here is that you have two sources of truth: the nodes in the etcd set, and the nodes in the leadership table. You want to reduce that to one source of truth (etcd) by simply stating that the nodes in the etcd are defacto leaders. They can (and probably should) delegate that to other nodes as needed (potentially via partitions), but getting those delegations should be done via the leaders.

Hope that helps, and good luck!

Reply:
Thanks for the write-up @withinboredom!

I think there is a bit of conflation with the leadership table and the Etcd cluster itself- I don't think the proposal makes that clear atm. Schedulers all share an nth of the jobs to roughly share the work across the replicaset, based on a `mod` of the random Job ID with the replica total being equal to the replica ID. The purpose of the leadership table to ensure consistency of job ingestion by both checking no two replicas work on the same nth partition, and that there is no partition overlap of work of the Job space by checking all replicas agree on the total observed. This partition leadership is part of the design of _Scheduler_ and would be true if the Etcd cluster was remote to the replica set (something we want to support), or if we were using a different database technology all together. The leadership table also allows us in future to support other use cases like having an active & passive replica set (assuming a remote Etcd cluster or similar). Since we already have this durable partition leadership that matches the lifecycle of each replica today, the design is suggesting we extend its use to also include the hostname port so it can be used in broadcasting the latest address list of the replica set.

The logic around handling scaling of embedded or non-embedded, or recovery of an embedded Etcd node is orthogonal to the leadership key space.

Reply:
In this case it is probably better to leave out the etcd stuff altogether and focus on the desired behavior and api this system is supposed to use. Leave the etcd stuff for a "default HA implementation" as theoretically it could be implemented by any database engine—even kubernetes configmaps and leases, probably.

Reply:
I've pushed a commit that will hopefully clear up the meaning here :) Thanks for the input.